song/song_machine.py

```diff
@@ -101,47 +100,54 @@ def update_state(self, category):
         else:
             # state change
             self.__lock = True
-            self.current_state = self.parser.states[next_state_name]
-
-            if self.current_state == self.last_state:
-                print("The END")
-
+            self.current_state = self.parser.song_parts[next_state_name]
             return True
 
 
 class SongParser:
-    NAME_STATES_TO_NOTES = "states_to_notes"
+    MAX_UTTERANCES = "max_num_utterances"
     NAME_CATEGORIES = "categories"
-    NAME_TRANSITIONS = "transitions"
-    NAME_FIRST_STATE = "first_state"
-    NAME_LAST_STATE = "last_state"
+    STATES = "states"
+    NAME_FIRST_PART = "first_part"
 
     data = {}
-    states = {}
+    song_parts = {}
     categories = []
 
     first_state_name = ''
-    last_state_name = ''
+    max_utterances = 0
 
     def __init__(self, validated_data):
         self.data = validated_data
 
     def parse(self):
         self.categories = self.data[self.NAME_CATEGORIES]
         self._create_states()
+        self._create_receipts()
         self._add_transitions()
 
     def _create_states(self):
-        for state_name, note in self.data[self.NAME_STATES_TO_NOTES].items():
-            self.states[state_name] = State(state_name, int(note))
-        self.first_state_name = self.data[self.NAME_FIRST_STATE]
-        self.last_state_name = self.data[self.NAME_LAST_STATE]
+        for part_dict in self.data[self.STATES]:
+            part_name = part_dict["name"]
+            note = part_dict["note"]
+            self.song_parts[part_name] = State(part_name, int(note))
+        self.first_state_name = self.data[self.NAME_FIRST_PART]
+
+    def _create_receipts(self):
+        for part_dict in self.data[self.STATES]:
+            part_name = part_dict["name"]
+            receipts = part_dict["receipts"]
+            self.song_parts[part_name].add_receipts(receipts)
```

Comment:
You iterate over `self.data[self.STATES]` again but have done it in `_create_states` already. I think I would remove the entire `_create_receipts` - function. Then add a `receipt`-argument to the constructor (`__init__()`) of `State` and add it in the creation of the instances in `_create_states`

Reply:
Done